mozilla-detekt-rules/src/main/java/org/mozilla/fenix/detektrules/MozillaStrictModeSuppression.kt

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.detektrules
+
+import io.gitlab.arturbosch.detekt.api.CodeSmell
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.api.Debt
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Rule
+import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtImportDirective
+
+private const val VIOLATION_MSG = "Please use `components.strictMode.resetAfter` instead because it has " +
+    "performance improvements and additional code to monitor for performance regressions."
+
+/**
+ * A check to prevent us from working around mechanisms we implemented to prevent suppressing StrictMode.
+ *
+ * IF YOU UPDATE THIS FILE NAME, UPDATE CODE OWNERS.
+ */
+class MozillaStrictModeSuppression(config: Config) : Rule(config) {
+    override val issue = Issue(
+        "MozillaStrictModeSuppression",
+        Severity.Performance,
+        "Prevents us from working around mechanisms we implemented to prevent suppressing StrictMode",
+        Debt.TEN_MINS
+    )
+
+    override fun visitImportDirective(importDirective: KtImportDirective) {
+        super.visitImportDirective(importDirective)
+        reportIfImportAcResetAfter(importDirective)
+    }
+
+    override fun visitCallExpression(expression: KtCallExpression) {
+        super.visitCallExpression(expression)
+        reportIfCallStrictModeSetPolicy(expression)
+    }
+
+    private fun reportIfImportAcResetAfter(importDirective: KtImportDirective) {
+        if (importDirective.importPath?.toString() == "mozilla.components.support.ktx.android.os.resetAfter") {
+            report(CodeSmell(issue, Entity.from(importDirective), VIOLATION_MSG))
+        }
+    }
+
+    private fun reportIfCallStrictModeSetPolicy(expression: KtCallExpression) {
+        // There is probably a more correct way of doing this but this API is not well documented
```

Comment:
@mcomella  what do you think about using ```override fun visitExpression(expression: KtExpression)```? I think this just goes through all the expressions in a file. So we could do something like : 

``` 
     if(expression.containingKtFile.name.contains("StrictModeManager")){
            if(expression.children.filter{ it.text == "setVmPolicy" 
                        || it.text ==   "setThreadPolicy"}.isNotEmpty()){ // make error here }
```

I'm not sure if its better, I just wasn't sure about looking for the receiver and caller method with the ```?``` .  Though, I only came up with this looking through the methods and playing with detekt.

Reply:
To address specific concerns I see with the example above,
1. With the visitor pattern, I think it's preferred to use the most specific node type that you can. If you use a more generic node type then you'll have to do that same filtering yourself and you can mess it up. For example, I believe your code would throw a violation on an assignment expression – `setVmPolicy = true`.
1. If you operate on all children rather than a specific child, then you could end up violating on the wrong expression. For example, in a hypothetical method, `strictModeManager.setVmPolicy.resetSuppressionCount()` – I believe your code would also violate here.

> I'm not sure if its better, I just wasn't sure about looking for the receiver and caller method with the ? . 

I don't think the goal should be to remove every optional type. Optional types are intended to express the absence of something and it's logical to use them in this capacity. The problem with optional types is that many programmers use them to express default values, edge case states, etc.

In this case, I think the optional types are used to express the absence of something. For example,
```
val receiver = expression.parent?
```

If I understand the API correctly, the callExpression means a function being invoked, while the parent is the receiver of that function. In some cases the receiver doesn't exist. For example, you can call `runBlocking { }`. Therefore, I think the comparison between `receiver == "StrictMode"` is fine given if there's no receiver, we're unlikely to be calling `setVmPolicy`.

I don't really understand why `calleeExpression` can be nullable though and that's a little concerning (maybe infix functions?). I'm not sure what to do about that though.

---

Your comment pointed out a bug in my code though – if setVmPolicy is imported directly, we won't catch it because I look for the StrictMode receiver. I'll add a comment about this and perhaps when we're adding a test we can go back and address this. I don't trust myself to write this correctly without tests (though I guess it could be an import statement block...).

Reply:
That makes sense to me. Its a bit difficult to know what's happening with the rule, that was another thing that was concerning. But I think that mostly has to do with Detekt APIs being so mysterious and lacking document, so I guess we can just add this part to our Detekt guide? I think the two points you mentioned are very valid and they aren't something I would've thought of. My first thought was "how can I get access to the ```StrictMode``` call?" which seems hard to do if you just look at the possible ```visit()``` methods and not all of them have a super clear name on what they do.